File: src/test/java/edu/ucdavis/fiehnlab/ctsRest/client/CtsClientTest.java
package edu.ucdavis.fiehnlab.ctsRest.client;

import edu.ucdavis.fiehnlab.config.CtsClientConfiguration;
import edu.ucdavis.fiehnlab.ctsRest.model.*;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.cloud.netflix.feign.EnableFeignClients;
import org.springframework.test.context.junit4.SpringRunner;

import java.io.UnsupportedEncodingException;
import java.util.List;
import java.util.Map;

import static org.junit.Assert.*;

/**
 * Created by diego on 2/16/2017.
 */
@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = CtsClientConfiguration.class)
@EnableFeignClients
public class CtsClientTest {
	private String inchikey = "MYMOFIZGZYHOMD-UHFFFAOYSA-N";
	private String inchiCode = "InChI=1S/O2/c1-2";
	private String molDef = "  2  1  0  0  0  0  0  0  0  0999 V2000\n" +
			"    0.0000    0.0000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\n" +
			"    1.2990   -0.7500    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\n" +
			"  2  1  2  0  0  0  0 \n" +
			"M  END";
	private String fullMol = " 977\\n" +
			"  -OEChem-07071413282D\\n\\n" +
			"  2  1  0  0  0  0  0  0  0  0999 V2000\\n" +
			"    0.0000    0.0000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\\n" +
			"    1.2990   -0.7500    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\\n" +
			"  2  1  2  0  0  0  0\\n" +
			"M  END";

    @LocalServerPort
    int port;

	@Autowired
	private CtsClient client;


	@Before
	public void setUp() {
		System.out.println("Client config: " + client.toString());
	}

	@After
	public void tearDown() {
		System.out.println("--------------------------------------------------------");
	}

	@Test
	public void testFromValues() {
		List<String> res = client.getSourceIdNames();

		assertNotNull(res);
		assertTrue(res.size() >= 10);
		assertTrue(res.contains("Chemical Name"));
		assertTrue(res.contains("Human Metabolome Database"));
		assertTrue(res.contains("InChIKey"));
		assertTrue(res.contains("KEGG"));
		assertTrue(res.contains("PubChem CID"));
		assertTrue(res.contains("ChemSpider"));
	}

	@Test
	public void testToValues() {
		List<String> res = client.getTargetIdNames();

		assertNotNull(res);
		assertTrue(res.size() >= 10);
		assertTrue(res.contains("Chemical Name"));
		assertTrue(res.contains("Human Metabolome Database"));
		assertTrue(res.contains("InChI Code"));
		assertTrue(res.contains("InChIKey"));
		assertTrue(res.contains("KEGG"));
		assertTrue(res.contains("PubChem CID"));
		assertTrue(res.contains("ChemSpider"));
	}

	@Test
	public void testChemifyQuery() {
		String res = client.chemifyQuery("alanine");

		assertNotNull(res);
	}

	@Test
	public void testBiologicalCount() {
		Map<String, Integer> res = client.compoundBiologicalCount(inchikey);

		assertNotNull(res);
		assertTrue(res.keySet().size() > 1);

		int sum = 0;
		for (Map.Entry<String, Integer> item : res.entrySet()) {
			assertTrue(item.getValue() > 0);
			if (!item.getKey().equals("total")) {
				sum += item.getValue();
			}
		}

		assertEquals(sum, res.get("total").intValue());
	}

	@Test
	public void testExtidCount() {
		ExtidCountResponse res = client.compoundExtidCount(inchikey);

		assertNotNull(res);
		assertTrue(res.datasource_count > 100);
	}

	@Test
	public void testFormulaExpansion() {
		FormulaResponse res = client.expandFormula("C8H10N4O2");

		assertNotNull(res);
		assertNull(res.error);
		assertEquals("CCCCCCCCHHHHHHHHHHNNNNOO", res.result);
	}

	@Test
	public void testCompoundInfo() {
		SynonymDto syn = new SynonymDto("Oxygen", "Synonym", 0);
		ExtidDto extid = new ExtidDto("PubChem CID", "977", "https://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi");

		CompoundResponse res = client.compoundInfo(inchikey);

		assertNotNull(res);
		assertEquals(inchikey, res.inchikey);
		assertTrue(res.synonyms.contains(syn));
		assertTrue(res.externalIds.contains(extid));
	}

	@Test
	public void testInchi2Mol() throws UnsupportedEncodingException {
		String inchicode = "{\"inchicode\":\"" + inchiCode + "\"}";

		MoleculeResponse res = client.inchi2Mol(inchicode);

		assertNotNull(res);
		assertNull(res.message);
		assertTrue(res.molecule.trim().endsWith(molDef.trim()));
	}

	@Test
	public void testInchikey2Mol() {
		MoleculeResponse res = client.inchiKey2Mol(inchikey);

		assertNotNull(res);
		assertNull(res.message);
		assertTrue(res.molecule.trim().endsWith(molDef.trim()));
	}

	@Test
	public void testMol2Inchi() {
		InchiPairResponse res = client.mol2Inchi("{\"mol\":\"" + fullMol + "\"}");

		assertNotNull(res);
		assertEquals(inchikey, res.inchikey);
		assertEquals(inchiCode, res.inchicode);
	}

	@Test
	public void testSmiles2Inchi() {
		String smiles = "{\"smiles\":\"[Cu+2].[O-]S(=O)(=O)[O-]\"}";

		String res = client.smiles2Inchi(smiles);

		assertNotNull(res);
		assertTrue(res.contains("\"inchicode\": \"InChI=1S/Cu.H2O4S/c;1-5(2,3)4/h;(H2,1,2,3,4)/q+2;/p-2\""));
	}

	@Test
	public void testInchiCode2InchiKey() {
		String inchicode = "{\"inchicode\":\"" + inchiCode + "\"}";

		Code2KeyResponse res = client.inchiCode2InchiKey(inchicode);

		assertNotNull(res);
		assertEquals(inchikey, res.inchikey);
	}

	@Test
	public void testConversion() {
		List<ConversionResult> res = client.convert("chemical name", "pubchem cid", "alanine");

		assertNotNull(res);
		assertTrue(res.size() > 0);
		assertArrayEquals(new String[]{"5950", "7311724", "57383916", "602", "71080", "7311725", "51283"}, res.get(0).results.toArray());
	}
}

File: ctsclient/src/test/scala/edu/ucdavis/fiehnlab/ctsrest/client/core/CtsClientTest.scala
package edu.ucdavis.fiehnlab.ctsrest.client.core

import com.typesafe.scalalogging.LazyLogging
import edu.ucdavis.fiehnlab.ctsrest.client.types.{InChIPairResponse, MoleculeResponse, ScoredInchi}
import edu.ucdavis.fiehnlab.ctsrest.casetojson.config.CaseClassToJSONSerializationAutoConfiguration
import org.junit.runner.RunWith
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.{Bean, Configuration}
import org.springframework.test.context.TestContextManager
import org.springframework.test.context.junit4.SpringRunner


@RunWith(classOf[SpringRunner])
@SpringBootTest(classes = Array(classOf[CtsClientTestConfig], classOf[CaseClassToJSONSerializationAutoConfiguration]))
class CtsClientTest extends AnyWordSpec with Matchers with LazyLogging {

  @Autowired
  val client: CtsClient = null

  new TestContextManager(this.getClass).prepareTestInstance(this)

  "CtsClientTest" should {
    val mol = MoleculeResponse("\n  CDK     1005182215\n\n  9  8  0  0  0  0  0  0  0  0999 V2000\n    1.2990   -0.7500    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    2.5981   -0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n    3.8971   -0.7500    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0\n    0.0000    0.0000    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0\n    0.3349   -1.8991    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0\n    2.2632   -1.8991    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0\n    3.5623    1.1491    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0\n    1.6339    1.1491    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0\n    5.1962   -0.0000    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0\n  2  1  1  0  0  0  0 \n  3  2  1  0  0  0  0 \n  1  4  1  0  0  0  0 \n  1  5  1  0  0  0  0 \n  1  6  1  0  0  0  0 \n  2  7  1  0  0  0  0 \n  2  8  1  0  0  0  0 \n  3  9  1  0  0  0  0 \nM  END\n", null)

    "have a client" in {
      client != null
    }

    "list source names" in {
      val sources = client.sourceIdNames()
      sources should not be empty
      sources should contain allOf ("PubChem CID","InChIKey","Chemical Name")
    }

    "list target names" in {
      val targets = client.targetIdNames()
      targets should not be empty
      targets should contain allOf ("PubChem CID","InChIKey","Chemical Name", "InChI Code")
    }

    "expandFormula" in {
      client.expandFormula("H2O") should have (
        'formula ("H2O"),
        'result ("HHO")
      )

      client.expandFormula("C2H6O") should have (
        'formula ("C2H6O"),
        'result ("CCHHHHHHO")
      )
    }

    "convert" in {
      val response = client.convert("chemical name", "inchikey", "alanine")

      response.head.fromIdentifier === "chemical name"
      response.head.searchTerm === "alanine"
      response.head.toIdentifier === "inchikey"
      response.head.results should contain allOf ("QNAYBMKLOCPYGJ-REOHCLBHSA-N", "QNAYBMKLOCPYGJ-UHFFFAOYSA-N", "QNAYBMKLOCPYGJ-UWTATZPHSA-N")
    }

    "score" in {
      val response = client.score("chemical name", "alanine", "biological")

      response.from should equal("chemical name")
      response.searchTerm should equal("alanine")
      response.result.head should equal(ScoredInchi("QNAYBMKLOCPYGJ-REOHCLBHSA-N", 1))
    }

    "compoundBiologicalCount" in {
      val response = client.compoundBiologicalCount("QNAYBMKLOCPYGJ-REOHCLBHSA-N")
      response should contain allOf ("KEGG" -> 2, "BioCyc" -> 1, "Human Metabolome Database" -> 1, "total" -> 4)
    }

    "compoundExtidCount" in {
      val response = client.compoundExtidCount("QNAYBMKLOCPYGJ-REOHCLBHSA-N")
      response.datasource_count === 277
    }

    "compoundSynonyms" in {
      val response = client.compoundSynonyms("QNAYBMKLOCPYGJ-REOHCLBHSA-N")
      response should contain allOf ("L-alanine", "Alanine", "(S)-2-amino-Propanoate")
    }

    "compoundInfo" in {
      val response = client.compoundInfo("QNAYBMKLOCPYGJ-REOHCLBHSA-N")

      response.inchikey === "QNAYBMKLOCPYGJ-REOHCLBHSA-N"
      response.inchicode === "InChI=1S/C3H7NO2/c1-2(4)3(5)6/h2H,4H2,1H3,(H,5,6)/t2-/m0/s1"
      response.exactmass === 89.09331
      response.molweight === 89.04768
      response.formula === "C3H7NO2"
    }

    "inchiKey2Mol" in {
      val response = client.inchiKey2Mol("LFQSCWFLJHTTHZ-UHFFFAOYSA-N")

      response.molecule.split("\n").filterNot(_.contains("  CDK  ")) should equal(mol.molecule.split("\n").filterNot(_.contains("  CDK  ")))
    }

    "inchi2Mol" in {
      var response = client.inchi2Mol("InChI=1S/C2H6O/c1-2-3/h3H,2H2,1H3")
      logger.info(s"RESPONSE: ${response}")

      // Drop header rows as CDK generated unique identifiers per conversion
      response.molecule.split('\n').drop(2) shouldEqual mol.molecule.split('\n').drop(2)
    }

    "mol2Inchi" in {
      val response = client.mol2Inchi(mol.molecule)
      logger.info(s"RESPONSE: ${response}")
      response shouldEqual InChIPairResponse("LFQSCWFLJHTTHZ-UHFFFAOYSA-N", "InChI=1S/C2H6O/c1-2-3/h3H,2H2,1H3")
    }

    "smiles2Inchi" in {
      val response = client.smiles2Inchi("CCO")
      logger.info(s"RESPONSE: ${response}")
      response.inchicode shouldEqual "InChI=1S/C2H6O/c1-2-3/h3H,2H2,1H3"
    }

    "inchiCode2InchiKey" in {
      val response = client.inchiCode2InchiKey("InChI=1S/C2H6O/c1-2-3/h3H,2H2,1H3")
      logger.info(s"RESPONSE: ${response}")
      response shouldEqual InChIPairResponse("LFQSCWFLJHTTHZ-UHFFFAOYSA-N", "InChI=1S/C2H6O/c1-2-3/h3H,2H2,1H3")
    }
  }
}

@Configuration
class CtsClientTestConfig {
  @Bean
  def client: CtsClient = new CtsClient
}
